src/conversions.rs

```diff
@@ -111,86 +228,195 @@ impl Optimizer {
+                    }
+                    None => Expr::Call("Void".into(), vec![]),
+                }],
+            ),
+            StructuredBlock::Basic(block) => {
+                Expr::Call("Basic".into(), vec![self.convert_basic_block(block)])
+            }
+        }
+    }
 
+    pub(crate) fn string_to_expr(&self, string: String) -> Expr {
+        Expr::Lit(egglog::ast::Literal::String(string.into()))
+    }
+
+    pub(crate) fn string_to_var_encoding(&self, string: String) -> Expr {
+        Expr::Call("Var".into(), vec![self.string_to_expr(string)])
+    }
+
+    /*pub(crate) fn string_to_var(&self, string: String) -> Expr {
+        Expr::Call("Var".into(), vec![self.string_to_expr(string)])
+    }*/
+
+    pub(crate) fn convert_basic_block(&mut self, block: &BasicBlock) -> Expr {
         // leave prints in order
         // leave any effects in order
         // leave assignments to variables used outside
         // of this function
         // otherwise inline
-        let mut res = Expr::Call("End".into(), vec![]);
         let mut env = HashMap::<String, Expr>::new();
-        for code in &func.instrs {
-            if let Code::Instruction(instr) = code {
-                self.add_instr_to_env(instr, &mut env);
-            }
-        }
-
-        // reverse order to build the linked list
-        for code in func.instrs.iter().rev() {
-            match code {
-                Code::Instruction(instr) => {
-                    res = self.add_instr_effect(instr, &res, &env);
-                }
-                Code::Label {
-                    pos: _pos,
-                    label: _label,
-                } => {
-                    // TODO handle labels
-                }
-            }
-        }
+        let codelist = block
+            .instrs
+            .iter()
+            .map(|instr| self.instr_to_code_expr(instr, &mut env))
+            .collect();
 
         Expr::Call(
-            "Func".into(),
+            "BlockNamed".into(),
             vec![
-                Expr::Lit(egglog::ast::Literal::String(func.name.clone().into())),
-                res,
+                Expr::Lit(egglog::ast::Literal::String(block.name.to_string().into())),
+                self.vec_to_codelist(codelist),
             ],
         )
     }
 
-    pub(crate) fn add_instr_effect(
+    pub(crate) fn instr_to_code_expr(
         &mut self,
         instr: &Instruction,
-        rest: &Expr,
-        env: &HashMap<String, Expr>,
+        env: &mut HashMap<String, Expr>,
     ) -> Expr {
-        match instr {
+        let (dest, expr) = match instr {
+            Instruction::Constant {
+                dest,
+                op: _op,
+                value,
+                pos: _pos,
+                const_type,
+            } => (
+                dest.clone(),
+                match value {
+                    Literal::Int(int) => Expr::Call(
```

Comment:
Why are these `Call` not `Lit`?

Reply:
We have to wrap them up so the egglog typechecker is happy